server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotsServiceNpeIT.java

```diff
@@ -0,0 +1,279 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class SnapshotsServiceNpeIT extends AbstractSnapshotIntegTestCase {
+
+    public void testNPE() throws Exception {
+        // 0 - Basic setup
+        final String masterNodeName = internalCluster().startNode();
+        final String dataNodeName = internalCluster().startDataOnlyNode();
+        createIndex(
+            "index-1",
+            Settings.builder().put("index.number_of_replicas", 0).put("index.routing.allocation.require._name", masterNodeName).build()
+        );
+        indexRandomDocs("index-1", 50);
+        createIndex(
+            "index-2",
+            Settings.builder().put("index.number_of_replicas", 0).put("index.routing.allocation.require._name", dataNodeName).build()
+        );
+        indexRandomDocs("index-2", 50);
+        createIndex(
+            "index-3",
+            Settings.builder().put("index.number_of_replicas", 0).put("index.routing.allocation.require._name", dataNodeName).build()
+        );
+        indexRandomDocs("index-3", 50);
+
+        // 1 - create repository and take a snapshot
+        final String repoName = "repo";
+        createRepository(repoName, TestRepositoryPlugin.REPO_TYPE);
+        final TestRepository testRepository = getRepositoryOnMaster(repoName);
+        logger.info("--> create snapshot snap-1");
+        createSnapshot(repoName, "snap-1", List.of("index-1"));
+
+        // 2 - Start deleting the snap-1 and block it at listing root blobs
+        PlainActionFuture<Void> future = setWaitForClusterState(state -> {
+            final SnapshotDeletionsInProgress snapshotDeletionsInProgress = SnapshotDeletionsInProgress.get(state);
+            return snapshotDeletionsInProgress.getEntries()
+                .stream()
+                .flatMap(entry -> entry.getSnapshots().stream())
+                .anyMatch(snapshotId -> snapshotId.getName().equals("snap-1"));
+
+        });
+        final CyclicBarrier barrier = testRepository.blockOnceForListBlobs();
+        new Thread(() -> {
+            logger.info("--> start deleting snapshot snap-1 ");
+            startDeleteSnapshot(repoName, "snap-1");
+        }).start();
+        assertBusy(() -> assertThat(barrier.getNumberWaiting(), equalTo(1)));
+        future.actionGet();
+        logger.info("--> repository blocked at listing root blobs");
+
+        // 3 - Stop data node so that index-2, index-3 become unassigned
+        internalCluster().stopNode(dataNodeName);
+        ensureColor(ClusterHealthStatus.RED, TimeValue.timeValueSeconds(30), false, "index-2", "index-3");
```

Comment:
I think `ensureColor(RED)` is confusing, this ensures that the health of these indices is _at least_ RED, which is always true. If you're wanting to wait for the data node to be removed then I think `validateClusterFormed()` or `ensureStableCluster()` would be best.

Reply:
Indeed this assertion is not really needed. I replaced it with `internalCluster().validateClusterFormed()`. Thanks!